proposals/filecoin-retrieval-in-ipfs.md

```diff
@@ -0,0 +1,163 @@
+
+High
+
+We know multiple companies that wish to use Filecoin as the storage provider for IPFS data. We know that many have been unsuccessful so far in bridging the gap.
+
+## Project definition
+#### Brief plan of attack
+
+1. We work with libp2p & IPFS stewards to come up with design for how we put filecoin retrieval records in the DHT. Note: I have used language so far that assumes these records will go in the IPFS DHT itself, but it may make sense to explore putting them in a seperate DHT that IPFS also has access too.
+
+2. We work out a way in Lotus to get these DHT records posted as part of the deal making process.
+
+3. We add the data transfer module to IPFS and the retrieval portion of go-fil-markets backed by a node interface that errors on chain based requests.
+
+4. We implement a negotiation mechanism for deciding between Filecoin retrieval via data transfer vs traditional bitswap in IPFS (note: this requires significant design). This negotiation happens in the [currently-in-development rewritten go-ipfs fetcher](https://github.com/ipfs/go-fetcher). Alternatively, perhaps data transfer itself becomes the top level interface above the fetcher? 
+
+5. We work with select pinning providers (eg Pinata) to verify we've provided the components for roundtrip pinning to Filecoin from IPFS
+
+6. We design the data transfer plugin API in concert with those who are looking at providing alternate mechanisms of retrieval (note: it's not clear to me if this should be a local plugin system or an HTTP API -- a local plugin can do any kind of custom HTTP communication, but is also harder to install and configure)
+
+<!--Briefly describe the milestones/steps/work needed for this project-->
+
+#### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+Done is when you can complete a roundtrip pinning / retrieval to Filecoin from IPFS and a spec exists for building future paid retrievals
+
+####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+- Well known pinning services begin offering Filecoin based pinning
+- Accelerator projects around new kinds of retrieval begin building plugins for paid retrieval
+
+<!--
+Provide success criteria. These might include particular metrics, desired changes in the types of bug reports being filed, desired changes in qualitative user feedback (measured via surveys, etc), etc.
+-->
+
+#### Counterpoints &amp; pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+- A DHT change discussion is a big design discussion, and could take a long time or hit roadblocks. It also could involve breaking changes to the content routing API in Libp2p, which might cause a lot of headaches
```

Comment:
Yes, depending on the implementation this may require some work on two fronts:
1. Deciding how we want to upgrade/bump the DHT protocol version. We have some ideas here, but AFAIK we've never done it. This is something worth doing anyway.
2. Making + testing the DHT changes